### ISSUE TITLE
pyjamas should not install files in /usr

### ISSUE THREAD

**Issue by pyjsorg** [pyjs/original-pyjs#576]
The pyjamas setup.py file has hard-coded references to /usr.  It should reference sys.prefix
instead.  Patch attached.  (I'm not a pyjamas user yet, so there is possibly more that needs to
be done.)

_Original issue: http://code.google.com/p/pyjamas/issues/detail?id=141 (March 12, 2009 02:41:30)_

**Comment by pyjsorg** [comment]
_From luke.lei...@gmail.com on March 12, 2009 11:28:44:_
they don't forcibly install files in /usr, that's just the default of your unix system.

use the --data-dir to over-ride your platform-specific data files location (used for
pyjamas libraries) which you may have forgotten to set, as _well_ as setting --prefix

**Comment by pyjsorg** [comment]
_From skip.mon...@gmail.com on March 12, 2009 12:05:47:_
As currently written I don't believe pyjamas can be installed using setuptools
(easy_install):

```
% easy_install-2.4 pyjamas
Searching for pyjamas
Reading http://pypi.python.org/simple/pyjamas/
Reading http://pyjs.org
Best match: Pyjamas 0.5
Downloading
```

http://pypi.python.org/packages/source/P/Pyjamas/Pyjamas-0.5.tar.gz#md5=f2cdc77dc168943efc88dd72c0475ab3
    Processing Pyjamas-0.5.tar.gz
    Running Pyjamas-0.5/setup.py -q bdist_egg --dist-dir
/tmp/easy_install-cP6Qv_/Pyjamas-0.5/egg-dist-tmp-qC-Czs
    error: Setup script exited with error: SandboxViolation:
mkdir('/usr/share/pyjamas',) {}

```
The package setup script has attempted to modify files on your system
that are not within the EasyInstall build area, and has been aborted.

This package cannot be safely installed by EasyInstall, and may not
support alternate installation locations even if you run its setup
script by hand.  Please inform the package's author and the EasyInstall
maintainers to find out if a fix or workaround is available.
```

Skip

**Comment by pyjsorg** [comment]
_From luke.lei...@gmail.com on March 14, 2009 14:35:43:_
skip, hi,

this is compiler technology, not "yet another python program".

the location of libraries _must_ go into a system-specific location, as defined by
FHS rules, on unix platforms.

that means: /usr/share.

developers who create their own pyjs-based applications will expect the pyjamas
libraries to go into /usr/share/pyjamas, and for upgrades to the compiler to replace
the libraries _not_ go and install somewhere completely different.

bottom line - you cannot use easy_install.

for windows users, tough luck on them: i've made the path self-relative.

**Comment by pyjsorg** [comment]
_From luke.lei...@gmail.com on March 14, 2009 14:40:21:_
skip, hi,

can you please raise a bugreport on http://bugs.python.org/setuptools/

you will need to mention that you are endeavouring to install a package which is
complying with FHS rules for linux by installing compiler libraries, to be used by
both the package and by other dependent packages, into /usr/share/pyjamas.

**Comment by pyjsorg** [comment]
_From luke.lei...@gmail.com on March 14, 2009 17:32:55:_
damn it, damn it.  lemme think about this one some more.

but definitely raise that bugreport with the setuptools developers.

**Comment by pyjsorg** [comment]
_From skip.mon...@gmail.com on March 14, 2009 17:51:44:_

> the location of libraries _must_ go into a system-specific location, as
> defined by FHS rules, on unix platforms.
> 
> that means: /usr/share.

That might be the perfect default place for it if you assume a standard
Linux packaging environment.  Let the folks who create Linux packages take
care of that.  If you install all pyjamas files relative to the root the
installer selected, you'll be fine.

I run on a Mac.  I have no idea if the filesystem hierarchy standard you
referred to applies there.  Even if it does though, as an individual I try
extremely hard _not_ to install stuff in /usr.  That's Apple's turf as far
as I'm concerned.  If I wanted to install a new version of gcc there's no
way in the world I would consider overwriting Apple's version.  I would set
--prefix to something like $HOME/local (as is my personal standard) when I
configured it and run gcc from ~/local/bin.  In fact, I have eight different
versions for Python installed on my machine, from 2.2 to 3.1.  None are
installed in such a way that a) conflict's with Apple's version or b)
requires me to su during installation.  At work we run Solaris and Linux and
never install Sun's Studio compiler, gcc or Python directly in /usr.

This is not a setuptools issue as far as I'm concerned so I won't be filing
a bug report there.  In my opinion setuptools is doing everything it needs
to do to give the person installing the package the option of establishing
an alternate root for the installation.  Applications, whether they are
compiler technology or not, need to respect the user's preference.

**Comment by pyjsorg** [comment]
_From nilton.v...@gmail.com on March 29, 2009 02:23:31:_
I stumbled upon this problem too. I don't think that hardcoding paths in a way that
does not allow the user to: 1) install the package without root password; 2) install
different versions of the same package; and 3) use easy_install; is a good thing.

**Comment by pyjsorg** [comment]
_From luke.lei...@gmail.com on April 12, 2009 13:11:32:_
ok - i've removed ezz setup from the equation.
ezz setup was fucking stupid and "decided it knew best", so it's been blown away.
by default, distutils installs stuff wherever the default location of the python
command is running.  that was the behaviour i was expecting - and the default
location on most unix systems is /usr.

the default location on mac systems is also /usr - if you use the default-installed
version of python.

regardless, you will now have to run "python bootstrap.py" and make the decision for
yourself.
